src/base/tag.h

```diff
@@ -0,0 +1,70 @@
+/*
+ * Bittorrent Client using Qt and libtorrent.
+ * Copyright (C) 2023  Vladimir Golovnev <glassez@yandex.ru>
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * In addition, as a special exception, the copyright holders give permission to
+ * link this program with the OpenSSL project's "OpenSSL" library (or with
+ * modified versions of it that use the same license as the "OpenSSL" library),
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#pragma once
+
+#include <compare>
+
+#include <QMetaType>
+#include <QString>
+
+class Tag final
+{
+public:
+    Tag() = default;
+
+    explicit Tag(const QString &tagStr);
+    explicit Tag(const std::string &tagStr);
+
+    bool isValid() const;
+    bool isEmpty() const;
+
+    QString toString() const noexcept;
+
+    bool operator==(const Tag &) const = default;
+    explicit operator QString() const noexcept;
```

Comment:
> Also it is kind of amazing that `QStringList` constructor is able to invoke this function, especially this function is marked `explicit`...

I don't really understand why are you confused. It works as intended to. `QList<T>` docs say:
>The value type of `InputIterator` must be convertible to `T`.

Reply:
> I think it would be valuable to add a new templated method `QList<T> OrderedSet::values<T>()`. Usage example: `m_tags.values<QString>();`.

Looks confusing as it usually happens when you try to combine several different actions in one (converting both container type and value type in this case). I refer to interface only, combining things can be done at the place of use.



> Of course I wish to introduce it myself and not delay this PR.

👍

Reply:
> Or it is you just not used to something new?

Is this something new? It looks like you want to do something that already exists (`QList(InputIterator first, InputIterator last)`). But your "new" way doesn't seem to have any advantages.

>I don't understand. Is the intention not well defined?

At least its proposed name `values` is inappropriate since it certainly does not express the intention, which is not to get a list of values of a set, but to convert them to another type.

>Doesn't programmer combine several operations together to become a meaningful function for reusing?

As I mentioned above, not everything is acceptable as part of something's interface, but it is quite acceptable outside of it.

But if we come back to this particular question, I don't see the point in reinventing the wheel (aka `QList<T> OrderedSet::values<T>()`), instead of using existing things (aka `QList(InputIterator first, InputIterator last)`).